eng/ci-tools/tools.go

```diff
@@ -4,9 +4,12 @@
 // Reference gotestsum to prevent it from being removed from go.mod by "go mod tidy".
 // See https://github.com/golang/go/wiki/Modules#how-can-i-track-tool-dependencies-for-a-module
 
+// This file fails to build if 'tools' is set, because gotestsum is a program, not a library. It is
+// necessary so "go install ..." works. There is no reason to build this file, anyway.
+
 //go:build tools
 // +build tools
 
 package ci_tools
 
-import _ "gotest.tools/gotestsum/cmd"
+import _ "gotest.tools/gotestsum"
```

Comment:
Interesting story here... with `import _ "gotest.tools/gotestsum"`, ` go build -tags tools ./...` fails:

```
tools.go:12:8: import "gotest.tools/gotestsum" is a program, not an importable package
```

I swapped it to `import _ "gotest.tools/gotestsum/cmd"`, because I thought it was reasonable that this file should be able to build. But that changes the dependencies, so after `go mod tidy`, `go install gotest.tools/gotestsum` fails:

```
C:\Users\dagood\go\pkg\mod\gotest.tools\gotestsum@v1.7.0\cmd\tool\slowest\ast.go:12:2:
  missing go.sum entry for module providing package golang.org/x/tools/go/packages (imported by gotest.tools/gotestsum/cmd/tool/slowest); to add:
        go get gotest.tools/gotestsum/cmd/tool/slowest@v1.7.0
```

So leaving it as an unbuildable `import _ "gotest.tools/gotestsum"` seems correct. There's no reason to build this file, after all. I've added a comment to sum this up.

Reply:
The errors you are seeing make sense. `gotest.tools/gotestsum` is a `main` package, so `go build`, `go test` and `go vet` will error out when seeing it in an import statement. That's why it is gated by a build constraint that should never be used. On the other hand, `go mod tidy` accepts `main` packages imports and does not take build constraints into account.

Swapping to `import _ "gotest.tools/gotestsum/cmd"` could be a valid move if you want to use any exported API from that package, which we don't 😄. Doing that would also make `gotest.tools/gotestsum` survive `go mod tidy`, as it is the root module of `gotest.tools/gotestsum/cmd`. On the other hand, this import will only add to our `go.sum` the dependencies used by `gotest.tools/gotestsum/cmd`, which does not include `golang.org/x/tools/go/packages` as it is only used in `gotest.tools/gotestsum/cmd/tool/slowest` and this package is not imported by the former.

Hopefully the tools story will be improved once https://github.com/golang/go/issues/48429 is approved and implemented.